src/bun.js/node/types.zig

```diff
@@ -1840,548 +1848,2863 @@ pub const Emitter = struct {
+    /// https://github.com/nodejs/node/blob/6ae20aa63de78294b18d5015481485b7cd8fbb60/lib/path.js#L919
+    /// The structs returned by parse methods.
+    fn ParsedPath(comptime T: type) type {
+        return struct {
+            root: []const T,
+            dir: []const T,
+            base: []const T,
+            ext: []const T,
+            name: []const T,
+        };
+    }
 
-        if ((ext.len != out.len or out.len == base_slice.len) and strings.endsWith(out, ext)) {
-            out = out[0 .. out.len - ext.len];
+    pub const MAX_PATH_BYTES = bun.MAX_PATH_BYTES;
+    pub const shim = Shimmer("Bun", "Path", @This());
+    pub const name = "Bun__Path";
+    pub const include = "Path.h";
+    pub const namespace = shim.namespace;
+    pub const sep_posix = CHAR_FORWARD_SLASH;
+    pub const sep_windows = CHAR_BACKWARD_SLASH;
+    pub const sep_str_posix = CHAR_STR_FORWARD_SLASH;
+    pub const sep_str_windows = CHAR_STR_BACKWARD_SLASH;
+
+    /// Taken from Zig 0.11.0 zig/src/resinator/rc.zig
+    /// https://github.com/ziglang/zig/blob/776cd673f206099012d789fd5d05d49dd72b9faa/src/resinator/rc.zig#L266
+    ///
+    /// Compares ASCII values case-insensitively, non-ASCII values are compared directly
+    fn eqlIgnoreCaseT(comptime T: type, a: []const T, b: []const T) bool {
+        if (@TypeOf(T) != u16) {
+            return std.ascii.eqlIgnoreCase(a, b);
+        }
+        if (a.len != b.len) return false;
+        for (a, b) |a_c, b_c| {
+            if (a_c < 128) {
+                if (std.ascii.toLower(@intCast(a_c)) != std.ascii.toLower(@intCast(b_c))) return false;
+            } else {
+                if (a_c != b_c) return false;
+            }
         }
+        return true;
+    }
 
-        return JSC.ZigString.init(out).withEncoding().toValueGC(globalThis);
+    /// Based on Node v21.6.1 private helper formatExt:
+    /// https://github.com/nodejs/node/blob/6ae20aa63de78294b18d5015481485b7cd8fbb60/lib/path.js#L130C10-L130C19
+    inline fn formatExtT(comptime T: type, ext: []const T, buf: []T) []const T {
+        const len = ext.len;
+        if (len == 0) {
+            return comptime L(T, "");
+        }
+        if (ext[0] == CHAR_DOT) {
+            return ext;
+        }
+        const bufSize = len + 1;
+        buf[0] = CHAR_DOT;
+        @memcpy(buf[1 ..bufSize], ext);
+        return buf[0 ..bufSize];
     }
 
-    fn dirnameWindows(path: []const u8) []const u8 {
-        if (path.len == 0)
-            return ".";
+    /// Caller must free returned memory.
+    fn getBufAlloc(allocator: std.mem.Allocator, len: usize) ![]u8 {
+        if (len > MAX_PATH_BYTES) {
+            const maybeBuf: ?[]u8 = allocator.alloc(u8, len) catch null;
+            const maybeTmpBuf1: ?[]u8 = allocator.alloc(u8, len) catch null;
+            const maybeTmpBuf2: ?[]u8 = allocator.alloc(u8, len) catch null;
+            const maybeTmpBuf3: ?[]u8 = allocator.alloc(u8, len) catch null;
```

Comment:
Do we free these buffers? I see we free the returned buf with `defer allocator.free(buf)` but not the temp bufs 1-3.

Reply:
The temp buffers we keep around as scratch buffers. But you're right. Encapsulated by the various T forms we could clear them and create them on call... I was keeping some scratch memory around for use to avoid allocating/freeing/allocating again.

Reply:
Made the tmp buffs freed.